Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
-        *listPtr = NSAtomicListSteal(sourceListPtr);
-        // source lists are in LIFO order, but we want to execute operations in the order they were enqueued
-        // so we reverse the list before we do anything with it
-        NSAtomicListReverse(listPtr);
-        op = NSAtomicListPop(listPtr);
+    unsigned int GetOperationCount() {
+        std::lock_guard<std::mutex> lock(_lock);
+
+        unsigned int ret = 0;
+
+        if (_curOperation != nil) {
+            ret++;
+        }
+
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            ret += _operationsHoldingArea[i].size();
+            ret += _candidateOperationsToRunNext[i].size();
+        }
+
+        return ret;
+    }
+
+    void CancelAllOperations() {
+        std::lock_guard<std::mutex> lock(_lock);
+
+        if (_curOperation != nil) {
+            [_curOperation cancel];
+        }
+
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            if (!_operationsHoldingArea[i].empty()) {
+                for (const auto& op : _operationsHoldingArea[i]) {
+                    [op cancel];
+                }
+            }
+
+            if (!_candidateOperationsToRunNext[i].empty()) {
+                for (const auto& op : _candidateOperationsToRunNext[i]) {
+                    [op cancel];
+                }
+            }
+        }
+    }
+
+    void Resume() {
+        BOOL shouldBroadcast = FALSE;
+        {
+            std::lock_guard<std::mutex> lock(_lock);
+            if (_isSuspended) {
+                _isSuspended = FALSE;
+                shouldBroadcast = TRUE;
+            }
+        }
+
+        if (shouldBroadcast) {
+            _suspendedCondition.notify_all();
+        }
     }
 
-    if (op != nil) {
-        if ([op isReady]) {
-            *curOperation = op;
-            [op start];
-            *curOperation = nil;
-        } else {
-            NSAtomicListInsert(sourceListPtr, op);
-            // We claim that we have done some work here.
-            // Otherwise, this work session is considered to be over and we clear the pending queue.
+    void Suspend() {
+        std::lock_guard<std::mutex> lock(_lock);
+        _isSuspended = TRUE;
+    }
+
+    BOOL IsSuspended() {
+        std::lock_guard<std::mutex> lock(_lock);
+        return _isSuspended;
+    }
+
+private:
+    // NOTE: this method is intended to be run under the lock but does not take the lock itself.
+    BOOL _HasMoreWorkNoLock() {
+        if (_curOperation != nil) {
             return TRUE;
         }
+
+        // NOTE: This may be insufficient / racy. This only checks the holding area and not "on deck" items.
+        // Its possible that the check is made right after a transfer to on deck which would incorrectly report
+        // all done when some items may go back into holding and certainly all of them aren't completed. This particularly
+        // would affect WaitUntilAllOperationsAreFinished and
```

Comment:
should be removed. was going to mention the thread loop but it doesn't actually impact that case.

---
In reply to: [95456923](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95456923) [](ancestors = 95456923)